### ISSUE TITLE
upgrade MKL-DNN 0.20.1 to DNNL 1.1

### ISSUE THREAD

**Issue by Jianhui-Li** [pytorch/pytorch#30300]
## 🚀 Feature
This RFC is to upgrade MKL-DNN 0.20.1 to DNNL 1.1. DNNL is MKL-DNN's new name.  The major release DNNL 1.0 introduced a number of new features, and DNN 1.1 continues improvement on performance, quality, and usability. Below is an incomplete list, please refer to https://github.com/intel/mkl-dnn/releases for a full description.   

    DNNL 1.0 added SGEMM copy-based kernels for Intel SSE 4.1, Intel AVX, Intel AVX 2 and Intel AVX 512 architectures. With this optimization Intel MKL-DNN’ JIT SGEMM implementation achieves comparable performance to Intel MKL.

    DNNL 1.0 introduced bfloat16 training and inference support in reorders, (de-)convolution, pooling, batch normalization, local response normalization, eltwise, inner product, shuffle, sum, and concat. The implementation relies on new instructions targeting future Intel Xeon Scalable processor (codename Cooper Lake). On Intel Xeon processors with Intel AVX512 support bfloat16 arithmetic is emulated.

    DNNL 1.0 introduced Intel Processor Graphics support covering fp16 and fp32 inference, and fp32 training. Intel MKL-DNN relies on OpenCL* runtime to execute computations on Intel Processor Graphics and provides interoperability with user’s OpenCL code.

    DNNL 1.1 includes (experimental) Introduced caching mechanism that reduces primitive creation time for repeated primitive creation. The functionality is disabled by default and has to be enabled in compile time.

## Additional context

DNNL 1.0 comes with a new API design, but the most change is absorbed by ideep, the DNNL integration layer.  Below is a list of changes to Pytorch and Caffe2 to enable the DNNL 1.1. 

		○ Build option change with new name (OMP:COMP vs. OMP)
		○ Implement ideep op using new DNNL operation (mul, add)
		○ Enable DNNL built-in primitive cache, remove hashing key to primitive cache in caffe2  
		○ change DNNL format representation (no explicit format tag like nchw). 
		○ Repalce passing allocator as op template with registering global allocator 
		○ Tensor construction code clean-up
		○ deconv IOHW workaround cleanup
		○ Simplify group conv interface

DNNL v1.1 PyTorch upgrade patch contains 3 components: 
	1. Upgrade underlying DNNL (original MKL-DNN) v0.21.1 to DNNL v1.1
	2. Upgrade DNNL integration layer (iDeep)
        3. Change build system and update PyTorch/Caffe2 integration

cc @VitalyFedyunin

**Comment by xsacha** [comment]
Should PyTorch integrate with the OpenCL support from this as well?

**Comment by Jianhui-Li** [comment]
@xsacha This upgrade doesn't intend to integrate OpenCL support to PyTorch. It upgrades to the DNNL 1.1 for CPU, not for GPU.

**Comment by gottbrath** [comment]
In the past MKL GEMMs have been faster than MKL-DNN gemms. The claim from Intel is that with DNNL 1.1 it will be faster ..ie no regressions in GEMMs.

**Comment by xsacha** [comment]
When I attempt to build MKL-DNN on Windows, it tells me that MSVC doesn't support OpenMP >= 3 required for MKL-DNN. Will this also be an issue on DNNL 1.1 implementation? That is, no Windows support again?

**Comment by vpirogov** [comment]
@xsacha,

Which version are you using? Intel MKL-DNN v0.21 and v1.0 build fine with MSVC, so if you are unable to build  the library on Windows it might be a bug.

**Comment by xsacha** [comment]
It is when I attempt to build it in Pytorch. With latest master, I get this output:
> -- MKL-DNN needs omp 3+ which is not supported in MSVC so far

**Comment by vpirogov** [comment]
I see now. This check is in [Pytorch's build system](https://github.com/pytorch/pytorch/blob/718feb6d76702a7f8898feaeb616a5a04a9acbad/cmake/Modules/FindMKLDNN.cmake#L21) and should probably be removed.

+@Jianhui-Li, @XiaobingSuper

**Comment by Jianhui-Li** [comment]
@vpirogov  Good catch. Will do.
